crates/wit-component/README.md

```diff
@@ -96,3 +96,85 @@ $ wasm-tools component wit demo.component.wasm
 
 Here the `demo.component.wasm` can now be shipped to a component runtime or
 embedded into hosts.
+
+## Custom Section Format
+
+Toolchains producing components typically have a workflow that looks something
+like this:
+
+1. A bindings generator, often based on [`wit-bindgen`], is used to generate
+  source code.
+2. A user writes source code against these bindings.
+3. The language's compiler runs and produces a core WebAssembly module.
+4. The core WebAssembly module is passed to this crate, producing a component.
+
+Steps (1) and (4) both take WIT as input and the goal is to make it such that
+you don't have to specify the same WIT on both ends. Instead the WIT from (1)
+should be seamlessly passed through to (4) without developers having to
+explicitly opt-in to it otherwise. To do this, the `wit-component` crate and
+componentization process uses a custom section.
+
+Part of the generated bindings from (1) above is source code (or similar) to
+instruct the language to embed a custom section in the final binary. The custom
+section's name must start with `component-type` but can have any number of
+characters following it (e.g. it must match the regex `/^component-type/`). A
+module is allowed to have a number of custom sections, including zero.
+
+Each custom section describes a WIT `world`. The unit of bindings generation is
+a `world` and each `world` used during step (1), which may possibly be in
+multiple separate libraries, will be encoded into custom sections. Custom
+sections typically have a name that includes the bindings generator,
+the bindings generator's version, the world, and an optional use-provided
+string. The contents of the custom section is a wasm-encoded WIT world.
+
+An example of this is that for this document:
+
+```wit
+package a:b;
+
+world foo {
+    import host: func();
+    export guest: func();
+}
+```
+
+this WIT world is encoded to wasm as:
+
+```wasm
+(component
+  (type (;0;)
+    (component
+      (type (;0;)
+        (component
+          (type (;0;) (func))
+          (import "host" (func (;0;) (type 0)))
+          (export (;1;) "guest" (func (type 0)))
+        )
+      )
+      (export (;0;) "a:b/foo" (component (type 0)))
+    )
+  )
+  (export (;1;) "foo" (type 0))
+  (@custom "package-docs" "\00{}")
+  (@producers
+    (processed-by "wit-component" "0.218.0")
+  )
+)
+```
+
+More details about the wasm encoding of WIT can [be found
+upstream](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md).
+
+Note here that the WIT world is just the single `component`-type export needed
+to decode a single world, no other items are encoded into this wasm.
+
+When determining the `world` for a wasm component the componentization process
```

Comment:
Perhaps we should mention the semver-aware part of this merge process here?

Reply:
Good point!